torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,715 @@
+  if (_rank == 0) {
+    for (auto socket : _masterSendingSockets) {
+      send_bytes<uint8_t>(socket,
+                          reinterpret_cast<uint8_t*>(ncclId),
+                          NCCL_UNIQUE_ID_BYTES);
+    }
+  } else {
+    recv_bytes<uint8_t>(_slaveSocket,
+                        reinterpret_cast<uint8_t*>(ncclId),
+                        NCCL_UNIQUE_ID_BYTES);
+  }
+}
+
+
+// Destructor will only close all the sockets
+DataChannelNccl::~DataChannelNccl() {
+   /**
+    * Note that destructor will be called after cudaruntime being unloaded since
+    * DataChannel is a global variable.
+    */
+  _destroySockets();
+}
+
+
+void DataChannelNccl::_destroySockets() {
+  // Destroying all the socket
+  if (_masterListeningSocket != -1) {
+    ::close(_masterListeningSocket);
+    _masterListeningSocket = -1;
+  }
+  if (_slaveSocket != -1) {
+    ::close(_slaveSocket);
+    _slaveSocket = -1;
+  }
+  for (size_t i = 0; i < _masterSendingSockets.size(); ++i) {
+    if (_masterSendingSockets[i] != -1) {
+      ::close(_masterSendingSockets[i]);
+      _masterSendingSockets[i] = -1;
+    }
+  }
+}
+
+// Destroy the data channel
+void DataChannelNccl::destroy() {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  // Destroying all the socket
+  _destroySockets();
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  /**
+   * Destroy the CUDA and NCCL resources
+   * TODO: creating C++ wrappers for CUDA and NCCL resources to do the
+   *       cleanup automatically
+   */
+  for (auto& itemPair : _groupNcclResources) {
+
+    auto groupId = itemPair.first;
+    auto devices = getDevicesList(_groupDevices[groupId]);
+
+    // Destroy the CUDA events
+    size_t idx = 0;
+    for (auto& event : *(itemPair.second.ncclCudaEvents())) {
+      gpuGuard.setDevice(devices[idx++]);
+      THCudaCheck(cudaEventSynchronize(event));
+      THCudaCheck(cudaEventDestroy(event));
+    }
+    // Destroy the communicators
+    for (auto& comm : *(itemPair.second.ncclComms())) {
+      NCCL_CHECK(ncclCommDestroy(comm));
+    }
+
+  }
+  _groupNcclResources.clear();
+  _groups.clear();
+  _groupDevices.clear();
+}
+
+
+// Destroy the resource for a single thread group
+void DataChannelNccl::destroyGroup(THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  if (_groupNcclResources.find(groupId) != _groupNcclResources.end()) {
+    // Guard GPU device
+    AutoGPU gpuGuard;
+    // Destroy the CUDA events
+    size_t idx = 0;
+    for (auto& event : *(_groupNcclResources[groupId].ncclCudaEvents())) {
+      auto devices = getDevicesList(_groupDevices[groupId]);
+      gpuGuard.setDevice(devices[idx++]);
+      THCudaCheck(cudaEventSynchronize(event));
+      THCudaCheck(cudaEventDestroy(event));
+    }
+    // Destroy the communicators
+    for (auto& comm : *(_groupNcclResources[groupId].ncclComms())) {
+      NCCL_CHECK(ncclCommDestroy(comm));
+    }
+    _groupNcclResources.erase(groupId);
+  }
```

Comment:
Would be nice if we could share this code with `destroy()`. Otherwise it's easy to miss something in one of these places when you'll change the other one